### PR TITLE
[WIP] Add a package: thrust.

### PR DESCRIPTION
**Pull request by weijianwen**

### CODE REVIEW

**Comment by ax3l** [comment]
careful, this also comes with the CUDA package: https://gist.github.com/ax3l/9489132

**Comment by weijianwen** [comment]
Thank you @ax3l . I didn't realize that standalone `thrust` library is so outdated. I'll confirm and close this PR if necessary.

**Comment by ax3l** [comment]
I think it's a great idea to add thrust but have no idea how spack handles the fact if an "artifact" such as trust is provided but an other package, too.

It's also a pitty that the main project organization seems on an internal server right now instead of being on github and public...

**Comment by ax3l** [comment]
Just to add on it: thrust with an OpenMP and TBB backend makes even sense without CUDA at all.

**Comment by alalazo** [comment]
Hi @weijianwen , what's the status of this PR?

**Comment by ax3l** [comment]
@alalazo can you advise how to handle the case that thrust is both provided (in a slightly older version) by this PR and by the `cuda` recipe?

**Comment by ax3l** [comment]
CUDa 9.0.69 (RC) is already shipping an "internal" thrust 1.9.0-patch4 without having it visible on github... aww...

**Comment by adamjstewart** [comment]
> can you advise how to handle the case that thrust is both provided (in a slightly older version) by this PR and by the `cuda` recipe (in its latest version)?

I don't think there's anything wrong with this. If someone wants a standalone `thrust` library, there's no reason not to package one. If there is any concern, we can add a comment to the description saying the `thrust` that comes with `cuda` is slightly newer.

File: var/spack/repos/builtin/packages/thrust/package.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    homepage = "https://thrust.github.io"
+    url      = "https://github.com/thrust/thrust/archive/1.8.2.tar.gz"
+
+    version('1.8.2', 'fc7fc807cba98640c816463b511fb53f')
+
+    def install(self, spec, prefix):
+        install_tree('doc', join_path(prefix, 'doc'))
+        install_tree('examples', join_path(prefix, 'examples'))
+        install_tree('thrust', join_path(prefix, 'include', 'thrust'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/thrust/package.py]: If you rebase on develop, you now have the ability to use `prefix.doc`, `prefix.examples`, and `prefix.include.thrust`.

**Comment by ax3l** [comment]
yes, but shouldn't this package expose versions for TBB, OpenMP and CUDA backends in thrust?

**Comment by adamjstewart** [comment]
If @weijianwen needs them, or is kind of enough to spend the time adding them, then yes. If not, anyone else can add it in a follow-up PR.

**Comment by ax3l** [comment]
As an update: `thrust` and `cub` versions are now developed in the open again and all matching tags to CUDA releases were pushed to GitHub :tada: 

https://github.com/thrust/thrust/issues/888#issuecomment-629872810

**Pull request merged**